src/content/en/fundamentals/media/fast-playback-with-video-preload.md

```diff
@@ -0,0 +1,571 @@
+
+Now, what if I want to speculatively pre-load some media content without
+knowing which piece of media user will eventually pick. If user is on a web
+page that contains 10 videos, we probably have enough memory to fetch one
+segment file from each but we should definitely not create 10 hidden video
+elements and 10 `MediaSource` objects and start feeding that data.
+
+The 2-parts example below shows you how to pre-cache multiple first segments of
+video using the powerful and easy-to-use Cache API. Note that something similar
+can be achieved with IndexedDB as well. We're not using service workers yet as
+the Cache API is also accessible from the Window object.
+
+#### Fetch and cache
+
+```
+const videoFileUrls = [
+  'bat_video_file_1.webm',
+  'cow_video_file_1.webm',
+  'dog_video_file_1.webm',
+  'fox_video_file_1.webm',
+];
+
+// Let's create a video pre-cache and store all first segments of videos inside.
+caches.open('video-pre-cache')
+.then(cache => Promise.all(videoFileUrls.map(videoFileUrl => fetchAndCache(videoFileUrl, cache))));
+
+function fetchAndCache(videoFileUrl, cache) {
+  // Check first if video is in the cache.
+  return cache.match(videoFileUrl)
+  .then(cacheResponse => {
+    // Let's return cached response if video is already in the cache.
+    if (cacheResponse) {
+      return cacheResponse;
+    }
+    // Otherwise, fetch the video from the network.
+    return fetch(videoFileUrl)
+    .then(networkResponse => {
+      // Add the response to the cache and return network response in parallel.
+      cache.put(videoFileUrl, networkResponse.clone());
+      return networkResponse;
+    });
+  });
+}
+```
+
+Note that if were to use HTTP Range request, I would have to recreate manually
+a `Response` object as the Cache API doesn't support Range responses [yet]. Be
+mindful that calling `networkResponse.arrayBuffer()` fetches the whole content
+of the response at once into renderer memory, hence why you may want to use
+small ranges.
+
+For reference, here's the modified part of the code above to save HTTP Range
+requests to the video pre-cache.
+
+```
+    ...
+    return fetch(videoFileUrl, { headers: { range: 'bytes=0-567139' } })
+    .then(networkResponse => networkResponse.arrayBuffer())
+    .then(data => {
+      const response = new Response(data);
+      // Add the response to the cache and return network response in parallel.
+      cache.put(videoFileUrl, response.clone());
+      return response;
+    });
+```
+
+#### Play video
+
+When user clicks a play button, we'll fetch first segment of video available in
+the Cache API so that playback starts immediately if available. Otherwise,
+we'll simply fetch it from the network. Keep in mind that browser and users may
+decide to clear the [Cache].
+
+As seen before, we use MSE to feed that first segment of video to the video
+element.
+
+```
+function onPlayButtonClick(videoFileUrl) {
+  video.load(); // Used to be able to play video later.
```

Comment:
See https://developers.google.com/web/updates/2017/06/play-request-was-interrupted and https://bugs.chromium.org/p/chromium/issues/detail?id=404161